modules/sugar/src/main/ts/ephox/sugar/api/node/ShadowDom.ts

```diff
@@ -0,0 +1,79 @@
+import {
+  Document, Element as DomElement,
+  HTMLElement, HTMLElementTagNameMap,
+  Node as DomNode,
+  ShadowRoot
+} from '@ephox/dom-globals';
+import * as Node from './Node';
+import * as Head from './Head';
+import * as Body from './Body';
+import { Option, Type } from '@ephox/katamari';
+import Element from './Element';
+import * as Traverse from '../search/Traverse';
+
+export type RootNode = Element<Document | ShadowRoot>;
+
+/**
+ * Is the element a ShadowRoot?
+ *
+ * Note: this is insufficient to test if any element is a shadow root, but it is sufficient to differentiate between
+ * a Document and a ShadowRoot.
+ */
+export const isShadowRoot = (dos: RootNode): dos is Element<ShadowRoot> =>
+  Node.isDocumentFragment(dos);
+
+/**
+ * Does the browser support shadow DOM?
+ *
+ * NOTE: Node.getRootNode() and Element.attachShadow don't exist on IE11 and pre-Chromium Edge.
+ */
+export const isSupported = (): boolean =>
+  Type.isFunction((DomElement.prototype as any).attachShadow) &&
+  Type.isFunction((DomNode.prototype as any).getRootNode);
+
+export const getRootNode = (e: Element<DomNode>): RootNode => {
+  if (isSupported()) {
+    return Element.fromDom((e.dom() as any).getRootNode());
+  } else {
+    return Traverse.documentOrOwner(e);
+  }
+};
+
+/**
+ * If this is a Document, return it.
+ * If this is a ShadowRoot, return its parent document.
+ */
```

Comment:
Missing some code? Or did it move and the comment was left behind?

Reply:
Ah. It moved. Fixed.